### PR TITLE
Fix getting changes for single commit branches

### PR DESCRIPTION
**Pull request by koeves**
- Fix getting changes for branches with one commit
- Put Acceptance back as valid target

----
 # ⚠️ Could not find ticket corresponding to `feature/fix_getting_changes. Does your branch name follow the correct pattern?

### CODE REVIEW

**Comment by koeves** [comment]
@SamTheisens do you know the reason for that change some time ago? Currently it breaks if you only have one commit in your branch, because then we diff the first commit with itself (the hex[0]...hex[-1]) part:
```
DEBUG    git.cmd  Popen(['git', 'diff', '--name-only', 'cdcf695d68066c949388cf95331b0663ff8227c2..cdcf695d68066c949388cf95331b0663ff8227c2'], 
```
https://jenkins.k8s-serv-backend.vdbinfra.nl/job/MPyL%20Pipeline/job/PR-10840/6/consoleFull (not sure if you have access to jenkins)

**Comment by SamTheisens** [comment]
@koeves that's because of this: https://github.com/Vandebron/mpyl/commit/9346192ebfe3f2b6b4f5e85e1c952b0bb04b4ce4. Even if main has never been checked out, we are still able to find out what happened on the current branch.
A fix for the "single (or no) commit on branch is in the PR here https://github.com/Vandebron/mpyl/pull/149/commits/1f1a880c0da4acabb52c877131445b78fc9bed44, besides a (very important) fix for making sure problems related to the repo aren't swallowed silently https://github.com/Vandebron/mpyl/pull/149/commits/ea351c5d7715d89bfaafd33a2acf49d6ea28a224

**Comment by koeves** [comment]
> @koeves that's because of this: [9346192](https://github.com/Vandebron/mpyl/commit/9346192ebfe3f2b6b4f5e85e1c952b0bb04b4ce4). Even if main has never been checked out, we are still able to find out what happened on the current branch. A fix for the _single (or no) commit on branch_ bug is in the PR here [1f1a880](https://github.com/Vandebron/mpyl/commit/1f1a880c0da4acabb52c877131445b78fc9bed44), besides a (very important) fix for making sure problems, like this one, related to the git repo aren't swallowed silently [ea351c5](https://github.com/Vandebron/mpyl/commit/ea351c5d7715d89bfaafd33a2acf49d6ea28a224)

Thanks, all clear now. I missed that part of the customizable stages PR then

**Comment by koeves** [comment]
I merged your PR and used a new version, but the single commit issue is still not fixed: https://jenkins.k8s-serv-backend.vdbinfra.nl/job/MPyL%20Pipeline/job/PR-10859/4
you can see that it still does: 
```
DEBUG    git.cmd  Popen(['git', 'rev-list', 'master..HEAD', '--'], cwd=/home/jenkins/agent/workspace/MPyL_Pipeline_PR-10859, universal_newlines=False, shell=None, istream=None)             
           DEBUG    git.cmd  Popen(['git', 'diff', '--name-only', '6756b00c85abd06ee629b9b30161cd8a5883eed6..6756b00c85abd06ee629b9b30161cd8a5883eed6']
```

i.e. getting the diff between the same revision 6756b00c85abd06ee629b9b30161cd8a5883eed6 (the first commit in my PR: https://github.com/Vandebron/Vandebron/pull/10859)

Also looking at the commit you linked, that seems to solve the empty branch issue but not the single commit one right? Since it checks if the revisions are None, not if it contains only one revision. I'm reopening this PR and I'll try another fix

**Comment by koeves** [comment]
New version works as well: https://jenkins.k8s-serv-backend.vdbinfra.nl/job/MPyL%20Pipeline/job/PR-10859/6

File: src/mpyl/schema/run_properties.schema.yml [hunk 1]
```diff
@@ -54,8 +54,8 @@ properties:
         properties:
           deploy_target:
             description: The deploy target.
-            type: ["string", "null"]
-            enum: [ null,  'PullRequest', 'PullRequestBase', 'Production' ]
+            type: [ "string", "null" ]
+            enum: [ null,  'PullRequest', 'PullRequestBase', 'Acceptance', 'Production' ]
```
Comment (koeves) [inline src/mpyl/schema/run_properties.schema.yml:58]: @SamTheisens Acceptance was taken out of the list in https://github.com/Vandebron/mpyl/commit/bf7e9f0535b4111288633daaf819976b3db9f133, I assume by accident?

Comment (koeves) [reply src/mpyl/schema/run_properties.schema.yml:58]: I should be more careful with reviewing I guess :D

File: src/mpyl/schema/run_properties.schema.yml [hunk 2]
```diff
@@ -1,29 +1,29 @@
 $schema: http://json-schema.org/draft-07/schema#
 $id: schema/run_properties.schema.yml
 required:
-  - build
+- build
```
Comment (Jorg88) [inline src/mpyl/schema/run_properties.schema.yml:4]: You seem to be doing this change a lot, should it be a formatting or linting rule if possible?

**Review by Jorg88 — APPROVED** [review]

File: src/mpyl/schema/run_properties.schema.yml (hunk 2 above)
Comment (koeves) [reply src/mpyl/schema/run_properties.schema.yml:4]: Yeah seems like my yaml formatter is different in pycharm. I'll try to disable it and revert these changes

File: src/mpyl/steps/models.py [hunk 3]
```diff
@@ -125,7 +125,7 @@ def from_configuration(run_properties: Dict, config: Dict):
             details=RunContext.from_configuration(build["run"]),
             target=Target(
                 build["parameters"].get("deploy_target", None)
-                or Target.PULL_REQUEST.value
+                or Target.PULL_REQUEST.value  # pylint: disable=E1101
```
Comment (Jorg88) [inline src/mpyl/steps/models.py:128]: These warnings are actually blocking all pr's, just ran into them in my pr after rebasing. I'll add a disable I guess, but the warnings actually seem valid when I look at the code? Maybe you can explain it to me when you're back

Comment (Jorg88) [reply src/mpyl/steps/models.py:128]: I fixed these in my cypress pr as well, using the `no-member` as Sam suggested. We do need to look into why these errors don't show up locally..

**Comment by koeves** [comment]
> If there is only one revision, `{revisions[0].hexsha}..{revisions[-1].hexsha}` are identical. So that should work. Did you try it out? Regarding https://github.com/Vandebron/mpyl/pull/155/files#diff-f2e93ed09e3ea127c21228d5487fed442fe1b3364117f6a7ec4705cd660a72a0R452, I think we should avoid using lint error codes because they're not readable. In this case we can replace `E1101` with `no-member` I think.

Exactly what I mean, git is trying to get the difference between the same two revisions (e.g. `git diff first_revision..first_revision`) which is empty! So if there's only one commit, mpyl detects no changes. I have reproduced this, and others have reported this as well.

You wrote above that with this approach we do not need to have master checked out. Is that a realistic use case we should cover? Also when we get the revisions in the same function we do `self._repo.iter_commits(f"{self._config.main_branch}..HEAD")` which already needs to have master.

I think we do need master checked out to correctly get the changes in a branch, otherwise what do you compare to? And also, master _is_ pulled by mpyl if it is not already there anyway.

So from my perspective I'd use the fix in this PR, or go back to the MPL version of doing just `git diff origin/master..`. 

Let me know what you think!

**Comment by SamTheisens** [comment]
Yes, makes sense. Doing `git diff origin/master...` gave problems after rebase IIRC. Don't remember exactly what, unfortunately.

**Comment by koeves** [comment]
> Yes, makes sense. Doing `git diff origin/master...` gave problems after rebase IIRC. Don't remember exactly what, unfortunately.

Thanks, then I'd merge the fix in this PR and see how it holds up over time

**Pull request merged**